pkg/planner/core/indexmerge_path.go

```diff
@@ -478,6 +631,183 @@ func (ds *DataSource) generateIndexMergeAndPaths(normalPathCnt int) *util.Access
+			if len(accessFilters) == 0 {
+				continue
+			}
+			paths, isIntersection, ok, err := ds.buildPartialPaths4MVIndex(accessFilters, idxCols, onePossibleMVIndexPath.Index)
+			if err != nil {
+				logutil.BgLogger().Debug("build index merge partial mv index paths failed", zap.Error(err))
+				return nil, nil, false, err
+			}
+			if isIntersection || !ok { // limitation 2
+				continue
+			}
+			if len(paths) == 0 {
+				continue
+			}
+			// UNION case, use the max count after access for simplicity.
+			maxCountAfterAccess := -1.0
+			for _, p := range paths {
+				maxCountAfterAccess = math.Max(maxCountAfterAccess, p.CountAfterAccess)
+			}
+			// Note that: here every path is about mv index path.
+			// find the most valuable mv index path, which means it has the minimum countAfterAccess.
+			if len(bestPaths) == 0 {
+				bestPaths = paths
+				bestCountAfterAccess = maxCountAfterAccess
+				bestNeedSelection = len(remainingFilters) != 0
+			} else if bestCountAfterAccess > maxCountAfterAccess {
+				bestPaths = paths
+				bestCountAfterAccess = maxCountAfterAccess
+				bestNeedSelection = len(remainingFilters) != 0
+			}
+		}
+		if len(bestPaths) != 0 {
+			usedMap[offset] = true
+			// correctly find a dnf condition for this mv index path
+			mvAndPartialPaths = append(mvAndPartialPaths, bestPaths...)
+			if !needSelection && bestNeedSelection {
+				// collect one path's need selection flag.
+				needSelection = bestNeedSelection
+			}
+		}
+	}
+	return mvAndPartialPaths, usedMap, needSelection, nil
+}
+
+// generateMVIndexMergePartialPaths4And try to find mv index merge partial path from a collection of cnf conditions.
+// step1: find all possible access mv index path, map these cnf conditions into it.
+//
+//	if we have two json predicate use a same mv index(c, j) path like: (1 member-of j) and (2 member-of j) and c=1
+//	even for this single mv index path, we should generate two partial index paths: and each from access mutations
+//		mutation 1: [c=1] and [1 member-of j]    ---> index partial path 1
+//		mutation 2: [c=1] and [2 member-of j]    ---> index partial path 2
+//	and finally merge this index partial path together, if there still some other json_col's predicate,
+//	build it partial index path out and merge it too.
+//
+// step2: after all mv index paths is traversed, try to find the normal index partial path. Let's say there is
+//
+//			index built on column ---> idx_c(c), since condition c is already used in previous mv index partial path,
+//			index idx_c is ignored. If there is a additional cnf item as d=1 and there is a idx_d(d) on it. Since
+//	     d=1 is not used in previous mv index partial path, we can build normal index partial path idx_d out from
+//			d=1.
+//
+// step3: after all mv index partial paths and normal index partial paths is all built. Build them up as a whole
+//
+//	index merge path together.
+//
+// this function mainly focus on step1.
```

Comment:
Then why do we put irrelevant comments here? I think comments should aim to make the code logic clear and should be easy to understand.
If some background information is needed, we can put that in outer functions like `generateIndexMerge4ComposedIndex`, and then reference those comments here.